allensdk/test/brain_observatory/test_session_api_utils.py

```diff
@@ -0,0 +1,227 @@
+    mixin_harness_fixture.set_params(**params_to_set)
+    obtained = mixin_harness_fixture.get_params()
+    assert obtained == expected
+
+
+@pytest.mark.parametrize("mixin_harness_fixture, params_to_set, expected", [
+    ({},
+     {'a_param': 5},
+     {'a_param_1': 8, 'a_param_2': 42.0, 'b_param_1': [1, 2, 3]}),
+
+    ({},
+     {'something_random': 10.0},
+     {'a_param_1': 8, 'a_param_2': 42.0, 'b_param_1': [1, 2, 3]}),
+
+], indirect=["mixin_harness_fixture"])
+def test_set_params_with_invalid_params(mixin_harness_fixture,
+                                        params_to_set, expected):
+    with warnings.catch_warnings(record=True) as w:
+        mixin_harness_fixture.set_params(**params_to_set)
+        assert 'not valid and is being ignored' in str(w[-1].message)
+
+    obtained = mixin_harness_fixture.get_params()
+    assert obtained == expected
+
+
+@pytest.mark.parametrize("mixin_harness_fixture, params_to_set, expected", [
+    ({},
+     {'a_param_1': 7.0},
+     {'a_param_1': 7, 'a_param_2': 42.0, 'b_param_1': [1, 2, 3]}),
+
+    ({},
+     {'a_param_2': '12'},
+     {'a_param_1': 8, 'a_param_2': 12.0, 'b_param_1': [1, 2, 3]}),
+
+    ({},
+     {'b_param_1': (5, 6, 7)},
+     {'a_param_1': 8, 'a_param_2': 42.0, 'b_param_1': [5, 6, 7]}),
+
+], indirect=["mixin_harness_fixture"])
+def test_set_params_typecasting_success(mixin_harness_fixture, params_to_set, expected):
+    with warnings.catch_warnings(record=True) as w:
+        mixin_harness_fixture.set_params(**params_to_set)
+        assert 'It was type-casted to' in str(w[-1].message)
+
+    obtained = mixin_harness_fixture.get_params()
+    assert obtained == expected
+
+
+@pytest.mark.parametrize("mixin_harness_fixture, params_to_set, expected", [
+    ({},
+     {'a_param_1': 'hello'},
+     {'a_param_1': 8, 'a_param_2': 42.0, 'b_param_1': [1, 2, 3]}),
+
+    ({},
+     {'a_param_2': [5]},
+     {'a_param_1': 8, 'a_param_2': 42.0, 'b_param_1': [1, 2, 3]}),
+
+    ({},
+     {'b_param_1': 1},
+     {'a_param_1': 8, 'a_param_2': 42.0, 'b_param_1': [1, 2, 3]}),
+
+], indirect=["mixin_harness_fixture"])
+def test_set_params_typecasting_failure(mixin_harness_fixture, params_to_set, expected):
+    with warnings.catch_warnings(record=True) as w:
+        mixin_harness_fixture.set_params(**params_to_set)
+        assert 'Type-casting failed' in str(w[-1].message)
+
+    obtained = mixin_harness_fixture.get_params()
+    assert obtained == expected
+
+
+@pytest.mark.parametrize("mixin_harness_fixture, params_to_set, data_params, expected", [
+    ({},
+     {'a_param_1': 42},
+     {'a_param_1'},
+     True),
+
+    ({},
+     {'a_param_1': 8},
+     {'a_param_1'},
+     False),
+
+    ({},
+     {'a_param_1': 8.0},
+     {'a_param_1'},
+     False),
+
+    ({},
+     {'a_param_2': 3.0},
+     {'a_param_1'},
+     False)
+
+], indirect=["mixin_harness_fixture"])
+def test_needs_data_refresh(mixin_harness_fixture, params_to_set, data_params, expected):
```

Comment:
Please add some tests for setting more than one parameter at once, and different types (bool, int, list, array, etc.)

Reply:
@kschelonka Many thanks for the suggestion, it definitely encountered some problems that I've hopefully addressed in the commits below!